src/mlpack/core/data/load_impl.hpp

```diff
@@ -77,51 +77,105 @@ bool inline inplace_transpose(arma::Mat<eT>& X)
   }
 }
 
-template<typename eT>
-bool Load(const std::string& filename,
-          arma::Mat<eT>& matrix,
-          const bool fatal,
-          const bool transpose)
+inline arma::file_type GuessFileType(std::istream& f)
 {
-  Timer::Start("loading_data");
+  // Taken from armadillo's function guess_file_type_internal
+  f.clear();
+  const std::fstream::pos_type pos1 = f.tellg();
 
-  // Get the extension.
-  std::string extension = Extension(filename);
+  f.clear();
+  f.seekg(0, std::ios::end);
 
-  // Catch nonexistent files by opening the stream ourselves.
-  std::fstream stream;
-#ifdef  _WIN32 // Always open in binary mode on Windows.
-  stream.open(filename.c_str(), std::fstream::in | std::fstream::binary);
-#else
-  stream.open(filename.c_str(), std::fstream::in);
-#endif
-  if (!stream.is_open())
+  f.clear();
+  const std::fstream::pos_type pos2 = f.tellg(); // pos2 holds length of stream
+
+  // Compute length of stream in N_max
+  const arma::uword N_max = ((pos1 >= 0) && (pos2 >= 0) && (pos2 > pos1)) ?
+      arma::uword(pos2 - pos1) : arma::uword(0);
+
+  f.clear();
+  f.seekg(pos1);
+
+  // Handle empty files.
+  if (N_max == 0)
+    return arma::file_type_unknown;
+
+  const arma::uword N_use = (std::min)(N_max, arma::uword(4096));
+
+  unsigned char* data_mem = (unsigned char*)
+      malloc(sizeof(unsigned char) * N_use);
```

Comment:
Please do not use `malloc()`. This is an abomination here.  Use `std::vector<unsigned char>` instead, which gives you automatic memory management (ie. it will clean up after itself).

https://en.cppreference.com/w/cpp/container/vector
http://www.cplusplus.com/reference/vector/vector/

You can get a pointer to the memory by taking the address of the first element in the vector:

    std::vector<unsigned char> data(N_use);
    unsigned char* data_mem = &data[0];